chaos-mesh/chaos-mesh
# Enable users to specify pod failure pause image via helm values

Signed-off-by: root <root@vm2.j5nfkxtvikcexczoernjzqy4zb.fx.internal.cloudapp.net>



### What problem does this PR solve?

Issue Number: close #1925

Problem Summary:

### What is changed and how it works?

Proposal: This PR allows users to specify a pod failure pause image via helm values. This is useful for for organizations who can only pull images from internal registries and also where organizations have policy applied where images can only come from xyz.com

What's Changed:

* Added a new environmental variable (controller deployment) for the pause image to the chaos controller
* Add this new environment variable to the controller config. Defaults to gcr.io pause image
* Add this controller config to the pod failure impl and set the pause image from this config


Tests



Manual Tests

Test 1
- Created a new Kind cluster as per development guide
- Build images and add images to kind nodes
```
./install.sh --local kind --dependency-only
```
- Install chaos mesh via helm 
```
./helm install chaosmesh chaos-mesh/helm/chaos-mesh/ --set controllerManager.podChaos.podFailure.pauseImage=abc.io/google-containers/pause:latest --set controllerManager.image=localhost:5000/pingcap/chaos-mesh
```
- Install sample nginx deployment
```
kind: Deployment
metadata:
  name: nginx-deployment
spec:
  selector:
    matchLabels:
      app: nginx
  replicas: 1
  template:
    metadata:
      labels:
        app: nginx
    spec:
      containers:
      - name: nginx
        image: nginx:1.14.2
        ports:
        - containerPort: 80
```
- Install Pod Failure 
```
apiVersion: chaos-mesh.org/v1alpha1
kind: PodChaos
metadata:
  name: pod-failure-example
spec:
  action: pod-failure
  mode: one
  value: ''
  selector:
    labelSelectors:
      'app': 'nginx'
```
- Get pod yaml and check if image is updated "abc.io/google-containers/pause:latest"
- Delete pod failure chaos and get image and reverted back to original

Test 2

- Repeat Test 1, but leave out controllerManager.podChaos.podFailure.pauseImage when installing helm. Get pod yaml of nginx pod and image should be updated to "gcr.io/google-containers/pause:latest"
- Delete pod failure chaos and get image and reverted back to original

Notes
- Needed to remove the crd for IoChaos installed via "./install.sh --local kind --dependency-only" as it had specified IoChaos, rather than expected IOChaos. Just applied from the crds directory
- In order to run unit tests and e2e tests via "make test", needed to move the location of kubebuilder to /usr/local/
```
cp -r /home/myhynes/chaos-mesh/output/bin/kubebuilder /usr/local/
```

### Release note 

```release-note
Please add a release note.
If you don't think this PR needs a release note then fill it with None.
```
